amiga: added support for Paula audio

BSzili opened the pull request:
Native Paula audio for Amiga plus optimized mixer for Paula 8-bit stereo and AHI 16-bit stereo.

sezero commented:
```
snd_paula.o(.text+0x1a): undefined reference to `custom'
snd_paula.o(.text+0x1f6): undefined reference to `custom'
snd_paula.o(.text+0x200): undefined reference to `custom'
snd_paula.o(.text+0x20c): undefined reference to `custom'
snd_paula.o(.text+0x270): undefined reference to `custom'
snd_paula.o(.text+0x276): more undefined references to `custom' follow
snd_paula.o(.text+0x2b8): undefined reference to `ciaa'
snd_paula.o(.text+0x32e): undefined reference to `ciaa'
snd_paula.o(.text+0x33e): undefined reference to `custom'
snd_paula.o(.text+0x346): undefined reference to `custom'
collect2: ld returned 1 exit status
make: *** [hexen2] Error 1
```

sezero commented:
Not in the error messages above, but I think `_volume` as referenced in asm should be `_sfxvolume`
(I had renamed that a very long time ago.)

sezero commented on engine/h2shared/snd_sys.h:
```diff
@@ -92,11 +93,13 @@
 
 #if defined(PLATFORM_AMIGA)
 #define HAVE_AHI_SOUND		1
+#define HAVE_PAULA_SOUND		1
```
These should be guarded by `PLATFORM_AMIGAOS3`, yes?

sezero replied:
Yes: MOS or AROS linkage fails, otherwise.

BSzili replied:
I forgot that PLATFORM_AMIGA is the common macro, I'll do that.

sezero commented:
> ```
> snd_paula.o(.text+0x1a): undefined reference to `custom'
> snd_paula.o(.text+0x1f6): undefined reference to `custom'
> snd_paula.o(.text+0x200): undefined reference to `custom'
> snd_paula.o(.text+0x20c): undefined reference to `custom'
> snd_paula.o(.text+0x270): undefined reference to `custom'
> snd_paula.o(.text+0x276): more undefined references to `custom' follow
> snd_paula.o(.text+0x2b8): undefined reference to `ciaa'
> snd_paula.o(.text+0x32e): undefined reference to `ciaa'
> snd_paula.o(.text+0x33e): undefined reference to `custom'
> snd_paula.o(.text+0x346): undefined reference to `custom'
> collect2: ld returned 1 exit status
> make: *** [hexen2] Error 1
> ```

AHI-Paula.lha::AHI-Paula/Src/14BitCal/Audio.c (from aminet)
I see:  `struct Custom *custom = (struct Custom*)0xdff000;`

No idea whether that is what we should do or not...

BSzili commented:
> Not in the error messages above, but I think `_volume` as referenced in asm should be `_sfxvolume` (I had renamed that a very long time ago.)

I removed the volume as this used to be part of the mixer in Quake where this code comes from:
```
  snd_vol = volume.value*256;
  (...)
  else if (shm->samplebits == 8)
  {
    unsigned char *out = (unsigned char *) pbuf;
    while (count--)
    {
      val = (*p * snd_vol) >> 8;
      p+= step;
      if (val > 0x7fff)
        val = 0x7fff;
      else if (val < (short)0x8000)
        val = (short)0x8000;
      out[out_idx] = (val>>8) + 128;
      out_idx = (out_idx + 1) & out_mask;
    }
```

The volume scaling was moved out of the mixer in Hexen 2, so I removed it from the assembly as well.

> AHI-Paula.lha::AHI-Paula/Src/14BitCal/Audio.c (from aminet) I see: `struct Custom *custom = (struct Custom*)0xdff000;`
> 
> No idea whether that is what we should do or not...

That can be done, but the ``custom`` and ``ciaa`` should be present in some link library. Is this with GCC 2?

sezero commented:
> That can be done, but the `custom` and `ciaa` should be present in some link library. Is this with GCC 2?

Yes, gcc2.95 (still don't have bebbo's toolchain.)

BSzili commented:
> > That can be done, but the `custom` and `ciaa` should be present in some link library. Is this with GCC 2?
> 
> Yes, gcc2.95 (still don't have bebbo's toolchain.)

I found them in the meantime, they are in libamiga.a. Is it present in that SDK? Maybe it needs manual linking, because it's not in the spec file?

sezero commented:
> > Yes, gcc2.95 (still don't have bebbo's toolchain.)
> 
> I found them in the meantime, they are in libamiga.a. Is it present in that SDK? Maybe it needs manual linking, because it's not in the spec file?

It is in the libnix version, but not the clib2 version, hmm..

BSzili commented:
That's unusual, it's should be part of libamiga.a, it's even mentioned in the official NDK headers. Oh well, I can get rid of the alib dependency.

sezero commented:
OK, these are in, thanks!

sezero commented:
One problem I detected by eyeballing is we don't have C-only fallback
for `(shm->signed8 &2)` case if one builds with `USE_M68K_ASM=no`:
We should either add one, or make the asm unconditional.
